### PR TITLE
Modify the FSS scheme to use Numpy

### PR DESCRIPTION
**Pull request by LaRiffle**
## Description
This improves on the previous implem of FSS, using numpy to have a faster runtime. The crypto store also reflects the changes. Tests are added and consolidated.

Fix: #3882

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/mpc/fss.py [hunk 1]
```diff
@@ -195,210 +226,365 @@ def le(x1, x2):
...
+            CWi = uncompress(_CW[i])
 
             for b in (0, 1):
-                τ = [g0, g1][b] ^ (t[i, b] * CW[i])
-                τ = τ.reshape(2, λ + 1, n_values)
-                # filtered_τ = τ[𝛼[i]] OLD
-                α_i = α[i].unsqueeze(0).expand(λ + 1, n_values).unsqueeze(0).long()
-                filtered_τ = th.gather(τ, 0, α_i).squeeze(0)
-                s[i + 1, b], t[i + 1, b] = split(filtered_τ, [λ, 1])
+                τ = [g0, g1][b] ^ (t[i, b] * CWi)
```
Comment (youben11) [inline syft/frameworks/torch/mpc/fss.py:256]: Is there a difference between those two t's, because it's kind of hard to spot the difference at first.

File: syft/frameworks/torch/mpc/fss.py [hunk 2]
```diff
@@ -195,210 +226,365 @@ def le(x1, x2):
...
+    # Numbers are on 64 bits
+    return buffer & 0b1111111111111111111111111111111111111111111111111111111111111110, buffer & 0b1
 
-        # Last tour, the other σ is also a leaf:
-        FnOutput[n] = t[n]
-        flat_result = FnOutput.sum(axis=0) % 2
-        return flat_result.reshape(original_shape)
+
+def huge_loop(seed_t_bytes, n_iter):
+    return [hashlib.sha3_256(seed_t_bytes[i * 16 : (i + 1) * 16]).digest() for i in range(n_iter)]
```
Comment (youben11) [inline syft/frameworks/torch/mpc/fss.py:442]: Wasn't shaloop purpose to reduce this computation?

File: syft/frameworks/torch/mpc/primitives.py [hunk 3]
```diff
@@ -13,90 +13,120 @@ class PrimitiveStorage:
...
+    def __init__(self, owner: AbstractWorker):
         """
         Their are below different kinds of primitives available.
         Each primitive stack is a fixed length list corresponding to all the
         components for the primitive. For example, the beaver triple primitive would
         have 3 components. Each component is a high dimensional tensor whose
         last dimension is the same and corresponds to the number of instances available
         for this primitive. That's why get_keys uses a quite complicated dimension
-        selector. This structure helps generating efficiently primitives using
+        selector. This structures helps generating efficiently primitives using
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/primitives.py:24]: TYPO: This-> Those

File: syft/frameworks/torch/mpc/primitives.py [hunk 4]
```diff
@@ -13,90 +13,120 @@ class PrimitiveStorage:
...
-                if remove:
-                    length = prim.shape[-1] - n_instances
-                    primitive_stack[i] = th.narrow(prim, -1, n_instances, length)
-
-            return keys
-        else:
-            raise EmptyCryptoPrimitiveStoreError(self, type_op, available_instances, n_instances)
+        primitive_stack = getattr(self, op)
+
+        if op in ("fss_eq", "fss_comp"):
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/primitives.py:54]: NIT: {"fss_eq", "fss_comp"} <-- to make it a set

File: syft/frameworks/torch/mpc/primitives.py [hunk 5]
```diff
@@ -13,90 +13,120 @@ class PrimitiveStorage:
...
+            available_instances = len(primitive_stack[0]) if len(primitive_stack) > 0 else -1
+            if available_instances >= n_instances:
+                keys = []
+                # We iterate on the different elements that constitute a given primitive, for
+                # example of the beaver triples, you would have 3 elements.
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/primitives.py:59]: I'll add them in the next PR, I wanted to make a reasonable PR first :)

File: syft/frameworks/torch/mpc/fss.py [hunk 6]
```diff
@@ -10,178 +10,209 @@
...
 import torch as th
 import syft as sy
+from syft.exceptions import EmptyCryptoPrimitiveStoreError
+from syft.workers.websocket_client import WebsocketClientWorker
+from syft.generic.utils import allow_command
+from syft.generic.utils import remote
 
 
-λ = 110  # 6  # 110 or 63  # security parameter
+λ = 127  # 6  # 110 or 63  # security parameter
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/fss.py:27]: Thanks, I'll change this!

File: syft/frameworks/torch/mpc/fss.py (hunk 2 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/fss.py:442]: Correct! This one is not used anymore... removing it.

File: syft/frameworks/torch/mpc/fss.py (hunk 1 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/fss.py:256]: Between t and τ (tau)?

Comment (LaRiffle) [reply syft/frameworks/torch/mpc/fss.py:256]: Or the two lines of code? (in which case `CW[i]` -> `CWi`

Comment (youben11) [reply syft/frameworks/torch/mpc/fss.py:255]: Between t and τ (tau), but I guess it make more sense when having the paper in hand and reading the code

Comment (LaRiffle) [reply syft/frameworks/torch/mpc/fss.py:255]: tau here is an intermediate variable (that I call `state` in the DIF btw, I'll rename it as such) which is used to select the first part or second part of the 2(λ + 1) bit-long `[g0, g1][b] ^ (t[i, b] * CWi)` depending of the bit `α[i]` .
Related to the paper: it consists of choosing the left branch or the right branch when descending the tree
